File: Block Sort/BlockSort.java
class BlockSort {

    static void insertionSort(int arr[], int left, int right) {

        for(int i = left + 1; i <= right; i++) {

            int key = arr[i];
            int j = i - 1;

            while(j >= left && arr[j] > key) {
                arr[j + 1] = arr[j];
                j--;
            }

            arr[j + 1] = key;
        }
    }

    static void blockSort(int arr[]) {

        int blockSize = 4;
        int n = arr.length;

        for(int i = 0; i < n; i += blockSize) {

            int right = Math.min(i + blockSize - 1, n - 1);

            insertionSort(arr, i, right);
        }

        for(int i = 1; i < n; i++) {

            int key = arr[i];
            int j = i - 1;

            while(j >= 0 && arr[j] > key) {
                arr[j + 1] = arr[j];
                j--;
            }

            arr[j + 1] = key;
        }
    }

    public static void main(String[] args) {

        int arr[] = {12, 4, 8, 15, 1, 9, 3, 7};

        blockSort(arr);

        for(int num : arr)
            System.out.print(num + " ");
    }
}

File: Block Sort/BlockSort.py
def insertion_sort(arr, left, right):

    for i in range(left + 1, right + 1):

        key = arr[i]
        j = i - 1

        while j >= left and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1

        arr[j + 1] = key

def block_sort(arr):

    block_size = 4
    n = len(arr)

    for i in range(0, n, block_size):

        right = min(i + block_size - 1, n - 1)

        insertion_sort(arr, i, right)

    for i in range(1, n):

        key = arr[i]
        j = i - 1

        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1

        arr[j + 1] = key

arr = [12, 4, 8, 15, 1, 9, 3, 7]

block_sort(arr)

print(arr)
